lib/core/src/Cardano/Wallet/Primitive/AddressDerivation/Icarus.hs

```diff
@@ -151,6 +172,84 @@ generateKeyFromSeed
     -> IcarusKey 'RootK XPrv
 generateKeyFromSeed = unsafeGenerateKeyFromSeed
 
+-- | Hardware Ledger devices generates keys from mnemonic using a different
+-- approach (different from the rest of Cardano).
+generateKeyFromHardwareLedger
+    :: Mnemonic 24
+        -- ^ The actual seed
+    -> Passphrase "encryption"
+        -- ^ Master encryption passphrase
+    -> IcarusKey 'RootK XPrv
+generateKeyFromHardwareLedger mnemonic (Passphrase pwd) = unsafeFromRight $ do
+    let seed = pbkdf2HmacSha512
+            $ T.encodeUtf8
+            $ T.intercalate " "
+            $ mnemonicToText mnemonic
+
+    let cc = hmacSha256 (BS.pack [1] <> seed)
+    let (kL, kR) = first normalizeEd25519 $ hashRepeatedly seed
+    pA <- ed25519ScalarMult kL
+
+    prv <- left show $ xprv $ kL <> kR <> pA <> cc
+    pure $ IcarusKey (xPrvChangePass (mempty :: ByteString) pwd prv)
+  where
+    unsafeFromRight :: (HasCallStack, Show e) => Either e a -> a
+    unsafeFromRight = either (error . show) id
+
+    hashRepeatedly :: ByteString -> (ByteString, ByteString)
+    hashRepeatedly bytes = case BS.splitAt 32 (hmacSha512 bytes) of
+        (kL, kR) | testBit (kL `BS.index` 31) 5 -> hashRepeatedly (kL <> kR)
+        (kL, kR) -> (kL, kR)
+
+    -- - Clear the lowest 3 bits of the first byte
+    -- - Clear the highest bit of the last byte
+    -- - Set the second highest bit of the last byte
+    --
+    -- As described in:
+    --
+    -- "BIP32-Ed25519 Hierarchical Deterministic Keys over a Non-linear Keyspace".
+    normalizeEd25519 :: ByteString -> ByteString
+    normalizeEd25519 bytes =
+        let
+            (firstByte, rest) = BS.splitAt 1 bytes
+            (rest', lastByte) = BS.splitAt 30 rest
+        in
+            mconcat
+                [ (firstByte `BS.index` 0)
+                    & (`clearBit` 0)
+                    & (`clearBit` 1)
+                    & (`clearBit` 2)
+                    & BS.singleton
+                , rest'
+                , lastByte `BS.index` 0
+                    & (`setBit` 6)
+                    & (`clearBit` 7)
+                    & BS.singleton
+                ]
+
+    ed25519ScalarMult :: ByteString -> Either String ByteString
+    ed25519ScalarMult bytes = do
+        scalar <- left show $ eitherCryptoError $ Ed25519.scalarDecodeLong bytes
+        pure $ Ed25519.pointEncode $ Ed25519.toPoint scalar
+
+    pbkdf2HmacSha512 :: ByteString -> ByteString
```

Comment:
Add a comment referencing [BIP39 From mnemonic to seed](https://github.com/bitcoin/bips/blob/master/bip-0039.mediawiki#From_mnemonic_to_seed).